plugins/alembic/module/vtkF3DAlembicReader.cxx

```diff
@@ -20,51 +22,338 @@
+    const IntermediateGeometry& originalData, IntermediateGeometry& duplicatedData)
+  {
+    bool needToDuplicate = originalData.uvFaceVarying || originalData.nIsFaceVarying;
+
+    auto uvMapIter = originalData.Attributes.find("uv");
+    auto nMapIter = originalData.Attributes.find("N");
+    bool haveUV = uvMapIter != originalData.Attributes.end();
+    bool haveN = nMapIter != originalData.Attributes.end();
+
+    if (needToDuplicate)
+    {
+      auto faceCount = originalData.Indices.size();
+      duplicatedData.Indices.resize(faceCount);
+      for (size_t i = 0; i < faceCount; i++)
+      {
+        auto thisFaceVertexCount = originalData.Indices[i].size();
+
+        // TODO check with nYue
+        duplicatedData.Indices[i].resize(
+          thisFaceVertexCount, Alembic::Abc::V3i(-9999, -9999, -9999));
+      }
+
+      // Points
+      {
+        V3fContainer pV3F;
+        int pRunningIndex = 0;
+        for (size_t i = 0; i < faceCount; i++)
+        {
+          auto thisFaceVertexCount = originalData.Indices[i].size();
+          for (size_t j = 0; j < thisFaceVertexCount; j++)
+          {
+            Alembic::Abc::V3f original_position =
+              originalData.Attributes.at("P")[originalData.Indices[i][j].x];
+            pV3F.emplace_back(original_position);
+            duplicatedData.Indices[i][j].x = pRunningIndex;
+            pRunningIndex++;
+          }
+        }
+
+        duplicatedData.Attributes.insert(AttributesContainer::value_type("P", pV3F));
+      }
+
+      // UV
+      if (haveUV)
+      {
+        V3fContainer uvV3F;
+        int uvRunningIndex = 0;
+
+        for (size_t i = 0; i < faceCount; i++)
+        {
+          auto thisFaceVertexCount = originalData.Indices[i].size();
+          for (size_t j = 0; j < thisFaceVertexCount; j++)
+          {
+            Alembic::Abc::V3f originalUV =
+              originalData.Attributes.at("uv")[originalData.Indices[i][j].y];
+            uvV3F.emplace_back(originalUV);
+            duplicatedData.Indices[i][j].y = uvRunningIndex;
+            uvRunningIndex++;
+          }
+        }
+
+        duplicatedData.Attributes.insert(AttributesContainer::value_type("uv", uvV3F));
+      }
+
+      // Normal
+      if (haveN)
+      {
+        V3fContainer nV3F;
+        int nRunningIndex = 0;
+
+        for (size_t i = 0; i < faceCount; i++)
+        {
+          auto thisFaceVertexCount = originalData.Indices[i].size();
+          for (size_t j = 0; j < thisFaceVertexCount; j++)
+          {
+            Alembic::Abc::V3f original_N =
+              originalData.Attributes.at("N")[originalData.Indices[i][j].z];
+            nV3F.emplace_back(original_N);
+            duplicatedData.Indices[i][j].z = nRunningIndex;
+            nRunningIndex++;
+          }
+        }
+
+        duplicatedData.Attributes.insert(AttributesContainer::value_type("N", nV3F));
+      }
+    }
+    else
+    {
+      duplicatedData = originalData;
+    }
```

Comment:
is there data somewhere that would test this code path @nyue ?

Reply:
[tetrahedron.zip](https://github.com/f3d-app/f3d/files/13396844/tetrahedron.zip)

I have tested it with the above a satisfy your testing requirements.

It's a tetrahedron. Both the UV and Normal are point attributes i.e. all non-facevarying